02-express-tutorial/app.js

```diff
@@ -1 +1,84 @@
-console.log('Express Tutorial')
+const express = require('express');
+const app = express();
+const { products } = require("./data")
+
+
+/**
+ * Create HTTP server
+ */
+// const http = require('http')
+const port = 3000;
+// app.set('port', port)
+// const server = http.createServer(app)
+// server.listen(port, () => {
+//     console.log(`app listening at http://localhost:${port}`)
+// })
+
+app.listen(port, () => {
+    console.log(`Local sercer is listening on port ${port}`)
+})
+
+app.use(express.static("./public"))
+
+// First test endpoint
+app.get('/api/v1/test', (request, response) => {
+    response.json({
+        message: 'My test API endpoint!'
+    })
+})
+
+// Get all products
+app.get('/api/v1/products', (request, response) => {
+    response.json(products)
+})
+
+// Get product by Id
+app.get('/api/v1/products/:productId', (request, response) => {
+    
+    const productId = parseInt(request.params.productId)
+    const product = products.find( p => p.id === productId )
+    
+    if( !product ) {
+        response.status(400).json({
+            message: 'Product not found!'
+        })
+    }
+
+    response.json(product)
+})
+
+// Search product query
+app.get('/api/v1/query', (request, response) => {
+    const defaultLimit = 5;
+
+    const search = request.query.search.toLocaleLowerCase()
+    const limit = parseInt(request.query.limit) || defaultLimit
+    const minPrice = parseInt(request.query.minPrice) || -1
+    const maxPrice = parseInt(request.query.maxPrice) || -1
+    
+    const resultProducts = products.filter(
+        (p) => {
+            if( search )
+                return p.name.toLowerCase().indexOf(search) >= 0
+                    && ( minPrice===-1 || p.price >= minPrice )
+                    && ( maxPrice===-1 || p.price <= maxPrice )
+            else
+                return false;
+        }
+    )
+
+    response.json({
```

Comment:
Typically on the resulting dataset is returned.  You don't have to send all of the query parameters back in the response.